---
title: "DMC_alltest Visualization"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#package setup

```{r}

#install.packages('ggforce')
library(RColorBrewer)
library(kableExtra)
library(readr)
library(tidyverse)
library(reshape2)
library(cowplot)
library(ggpubr)
library(egg)
library(ggplot2)
library(cowplot)
library(ggforce)
```



#data setup
```{r}
# parent directory
base_path <- file.path("Z:", "Victor", "Projects", "NIH - cells", "data", "main", "interim")

# sample data
pDat <- readRDS(file.path(base_path, '3_1_pDat_filt.rds'))

# methylation data
betas <- readRDS(file.path(base_path, '3_1_betas_filt.rds'))

# annotation
anno <- readRDS(file.path('Z:', 'Victor', 'Repositories', 'EPIC_annotation', 'hg19_epic_annotation.rds'))

#color code
color_code <- readRDS(file.path(base_path, '2_3_color_code.rds'))
color_code_tissue <- setNames(color_code$Colors_Tissue, color_code$label)
colors <- color_code_tissue[unique(pDat$Tissue)]

#pull out gene name for search list use
gene <- tibble(gene = str_split(anno$genes_symbol, ', ') %>%
        unlist() %>%
         unique()) %>%
  filter(!is.na(gene))
#remove NA
gene <- gene[-9,]
#pull out cpg name for search list use
cpg <- unique(anno$cpg) 
cpg <- cpg[cpg %in% rownames(betas)] %>% as.tibble()

```


#beta boxplot
##Main function1(produce a single plot when clicking a cpg site name)
```{r}
#boxplots which show beta values based on cell types
boxplot_individual<- function(cpg_name){
  
    #get sample cell types and trimester
    cpg <- sample_info_cpg(cpg_name)
    
    #boxplot
    boxplot_trimester(cpg)
}

```

##Main function2(produce boxplots when having multiple cpg sites)
```{r}
#boxplot to show different tissue types
boxplot_total <- function(cpg_name = NULL, gene_name = NULL, nrow = 3){
  #when only have gene inputs
  if(is.null(cpg_name)){
  cpg <- sample_info_gene(gene_name)
  }
  
  #when only have cpg inputs
  else if(gene_name ==''){
  cpg <- sample_info_cpg(cpg_name)  
  }
  
  #when have both cpg and gene input
  else{
  stop('please enter either cpgs or a gene')
  }
  #boxplot
  boxplot_tissue(cpg, nrow)
}

```

##pull out cpg information(beta, tissue and trimester)
```{r}
#get sample information (beta, tissue and trimester)
sample_info_cpg <- function(cpg_name){
    
    cpg <- betas[cpg_name,]
    #when have only one cpg input
    if (length(cpg_name) ==1){
    cpg <- as.data.frame(cpg) %>% t()
    cpg <- cpg[,colnames(cpg) %in% pDat$Sample_Name]
    cpg <- as.data.frame(cpg)
    colnames(cpg) <- cpg_name
    cpg <- cpg %>% cbind(pDat) %>% gather(cpg, betas, contains('cg'))
    }
    #when have more than one cpg inputs
    else{
    cpg <- cpg[,colnames(cpg) %in% pDat$Sample_Name]
    cpg <- cpg %>% t() %>% cbind(pDat) %>% gather(cpg, betas, contains('cg'))
    }
    cpg <- cpg %>% left_join(anno, by = 'cpg') %>% arrange(desc(start)) %>% mutate(cpg = factor(cpg, levels = unique(cpg)))
    cpg
}

#information for producing boxplot which shows beta values of cpg sites related input genes
sample_info_gene <- function(gene_name){
  #get related cpg sites
  cpg_name <- fn_cpg(gene_name)
  #get sample cell types and tri
  cpg_info <- sample_info_cpg(cpg_name) 
  cpg_info
}


```

###get related cpg sites
```{r}
fn_cpg <- function(gene_name){
  cpg_name <- anno[grepl(gene_name, anno$genes_symbol),]
  cpg_name <- cpg_name$cpg %>% unique()
  cpg_name <-  cpg_name[cpg_name %in% rownames(betas)]
  
  cpg_name
}

```

##ggplot function
```{r}
#boxplot function
boxplot_tissue <- function(cpg, nrow = 3){
  if (length(unique(cpg$cpg)) >7)
  {ggplot(cpg, aes(x=Tissue, y=betas, fill = Tissue)) +
      geom_violin()+
      geom_jitter(aes(color = Tissue), alpha = 0.5, show.legend = F) +
      facet_wrap_paginate(.~cpg, ncol = 7, nrow = nrow)+ 
      theme_bw()+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust =1))+
      labs(x = ' ', y ='betas') +scale_fill_manual(values = colors) + scale_color_manual(values = colors) + ylim(0,1)
  }
  else
  {
    ggplot(cpg, aes(x=Tissue, y=betas, fill = Tissue)) +
      geom_violin()+
      geom_jitter(aes(color = Tissue), alpha = 0.5, show.legend = F) +
      facet_wrap_paginate(.~cpg)+ 
      theme_bw()+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust =1))+
      labs(x = ' ', y ='betas') +scale_fill_manual(values = colors) + scale_color_manual(values = colors) + ylim(0,1)
  }
}

boxplot_trimester <- function(cpg){
  

  ggplot(cpg, aes(x=Tissue, y=betas, fill = Tissue)) +
      geom_violin()+
      geom_jitter(aes(color = Tissue), alpha = 0.5, show.legend = F) +
      facet_wrap_paginate(.~Trimester)+ 
      theme_bw()+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust =1))+
      labs(x = ' ', y ='betas') +scale_fill_manual(values = colors) + scale_color_manual(values = colors) + ylim(0,1)
  
}

```


#datatable which shows cpg annotation
##Main function
```{r}
#datatable which shows cpg annotation (when inputs have both cpgs and genes)
cpg_info_total <- function(cpg_name = NULL, gene_name = NULL){
    if (is.null(cpg_name)){
      info <- cpg_info_gene(gene_name)
    }
    else if (gene_name == ''){
      info <- cpg_info(cpg_name)
    }
    else {
      stop('please enter either cpgs or a gene')
    }
    
  info
}
```

##when input only cpgs
```{r}
#datatable which shows cpg annotation
anno1 <- anno %>% select(cpg,chr,start,end, cpg_id, enhancers_id, genes_id, genes_symbol, genes_tx_id, pmd_id, imprinted_dmr_general, imprinted_dmr_placenta)

cpg_info <- function(cpg_name){

  cpg <- filter(anno1, cpg %in% cpg_name) %>% arrange(desc(start))

cpg
}
```

##when input only genes
```{r}
#datatable shows all cpg sites related to input gene
cpg_info_gene <- function(gene_name){

  #get cpgs sites related to input genes
  cpg_name <- fn_cpg(gene_name)
  
  cpg <- filter(anno1, cpg %in% cpg_name)%>% arrange(desc(start))
  cpg
}


```


#plots to view mean, sd and annotation
##main function
```{r}

stat_plot <- function(cpg_name = NULL, gene_name = NULL){
   if (is.null(cpg_name)){
      plot <- plot_cpgs_in_gene(gene_name)
    }
    else if (gene_name == ''){
      plot <- plot_cpgs(cpg_name)
    }
    else {
      stop('please enter either cpgs or a gene')
    }
  plot
}
```

##gene
```{r}
plot_cpgs_in_gene <- function(gene, font_size = 12, point_size = 1) {
  # gene is the gene symbol to match plot
  # betas is the methylation data
  # pDat is the sample information which includes important group info to plot by
  # annotation is links cpgs to genes and other cpg annotation to plot

  # filter to gene in annotation
  anno_gene <- anno %>%
    filter(grepl(paste0('\\<', gene, '\\>'), genes_symbol),
           cpg %in% rownames(betas)) %>%
    arrange(start)
  
  if (nrow(anno_gene) == 0) {
    stop('No cpgs found, maybe they were filtered out.')
  }
  
  # generate mean and standard deviation of each cell type / trimester
  betas_gene <- t(betas[anno_gene$cpg,]*100) %>% as_tibble %>% 
    mutate(Sample_Name = colnames(betas)) %>%
    
    # reshape into longer format
    pivot_longer(cols = -Sample_Name,
                 names_to = 'cpg', 
                 values_to = 'beta')  %>%
    
    # add tissue and trimester info
    left_join(pDat %>% select(Sample_Name, Trimester, Tissue), by = 'Sample_Name') %>%
    
    # calculate mean and sd for each cpg for each group
    group_by(Tissue, Trimester, cpg) %>%
    summarize(mean = mean(beta),
              sd = sd(beta)) %>%
    mutate(lower = mean-sd, upper = mean+sd) %>%
    
    # add cpg info
    left_join(anno_gene, by = 'cpg') %>%
    
    # order on cpg position
    ungroup() %>%
    arrange(desc(start)) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           Trimester_Tissue = paste0(Trimester, ' - ', Tissue))
  
  # Generate  methylation track
  p1 <- betas_gene %>%
    ggplot() +
    
    # geom_ribbon is to generate the alternating shaded background
    geom_ribbon(aes(x = as.numeric(cpg), ymin = 0, ymax = 25),
                fill = 'grey', alpha = 0.25)+
    geom_ribbon(aes(x = as.numeric(cpg), ymin = 50, ymax = 75),
                fill = 'grey', alpha = 0.25)+
    
    # geom_linerange and geom_point is for the actual methylation data
    geom_linerange(alpha = 0.5, size = point_size, 
                   aes(x = cpg, ymin =lower, ymax = upper, color = Tissue)) +
    geom_point(alpha = 0.75, aes(x = cpg, y = mean, color = Tissue), size = point_size*2) +
    
    # we want to facet by trimester
    facet_grid(rows = vars(Trimester)) +
    
    # cosmetics
    theme_bw(base_size = font_size) +
    theme(axis.text.x = element_blank(), 
          axis.ticks.x = element_blank(),
          panel.grid.minor.y = element_blank(),
          panel.grid.major.y = element_blank(),
          plot.margin=margin(l=-0.1,unit="cm"),
          axis.title.y = element_text(angle = 0, vjust = 0.5)) +
    scale_y_continuous(limits = c(0, 100), expand = c(0, 0)) +
    scale_x_discrete(expand = c(0.01,0.01))+
    scale_color_manual(values= color_code_tissue[unique(betas_gene$Tissue)]) +
    labs(y = '% methylation', x = '', color = '', title = gene) 
    
  # plot cpg annotations
  p2 <- betas_gene %>% 
    
    # First we need to process the annotation data
    mutate(# absent/presence for different genomic elements
           enhancer = !is.na(enhancers_id),
           pmd = !is.na(pmd_id),
           imprinted_gene_placenta = !is.na(imprinted_gene_placenta),
           imprinted_gene_general = !is.na(imprinted_gene_general),
           imprinted_dmr_general = !is.na(imprinted_dmr_general),
           imprinted_dmr_placenta = !is.na(imprinted_dmr_placenta)) %>%
    mutate_if(is.logical, as.character) %>%
    select(cpg, enhancer, pmd, imprinted_gene_general:imprinted_dmr_placenta, cpg_id, start) %>%
    
    # reshape
    pivot_longer(cols = c(-cpg, -start),
               names_to = 'cpg_element', 
               values_to = 'presence') %>%
    distinct() %>%
    
    # arrange plot orde for each track
    arrange(desc(start)) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           cpg_element = factor(cpg_element, 
                                levels = c('enhancer', 'pmd', 'imprinted_gene_placenta',
                                           'imprinted_gene_general', 'imprinted_dmr_general',
                                           'imprinted_dmr_placenta', 'cpg_id')),
           presence = factor(ifelse(presence == "TRUE", 'Present',
                                    ifelse(presence == "FALSE", 'Absent', presence)),
                             levels = c('Present', 'Absent', ' ', '  ', 
                                        'sea', 'shelf', 'shore', 'island')),
           # for facet label
           group = 'Annotations') %>%
    
    # plot code
    ggplot(aes(x = cpg, y = cpg_element, fill = presence)) +
    scale_fill_manual(values = c('Present' = '#cccccc', 'Absent' = '#f7f7f7', 
                                 ' ' = 'white', '  ' = 'white',
                                 'sea' = '#ffffcc', 'shelf' = '#a1dab4', 'shore' = '#41b6c4',
                                 'island' = '#225ea8'), drop = T) +
    geom_tile(color = '#969696') +
    guides(fill=guide_legend(ncol=3, override.aes=list(colour=NA))) +
    facet_grid(group~.) +
    theme_bw(base_size = font_size) +
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          legend.key = element_rect(colour = NA), 
          plot.margin = margin(l=-0.1,unit="cm")) +
    scale_y_discrete(expand = c(0,0)) +
    scale_x_discrete(expand = c(0,0)) +
    labs(x = '', y = '', fill = '')
  
  # transcript annotations
  
  # cpgs can have more than one mapping to a gene element per transcript
  # cpgs mapped to intronexonboundaries can also be part of exons and introns
  # 5 UTR and 3 UTR can also overlap with exons and introns
  
  # If there are multiple mappings then only the following will be displayed in order of priority:
  # 5UTR / 3UTR, exon / intron, intronexonboundary
  
  p3 <- betas_gene %>% 
    select(cpg, end, genes_id, genes_tx_id) %>%
    distinct() %>%
    
    # split comma separate list and put each entry into a new row
    mutate(genes_tx_id = str_split(genes_tx_id, ', '),
           genes_id = str_split(genes_id, ', ')) %>%
    unnest(c(genes_tx_id, genes_id)) %>%
    
    # order transcripts by position of start
    arrange(end) %>%
    mutate(genes_tx_id = factor(genes_tx_id, levels = unique(genes_tx_id))) %>%
    
    # fill all combinations of cpg and gene_tx_id with NA if missing, so that they show in plot
    complete(cpg, genes_tx_id) %>%
    
    # filter out to unique mappings using the priority described above
    # strategy is to order by factor level and take the first occurence
    mutate(genes_id = factor(genes_id, levels = c('1to5kb', 'promoter', '5UTR', '3UTR', 'exon', 
                                                  'intron', 'intronexonboundary', 'intergenic'))) %>%
    
    group_by(cpg, genes_tx_id) %>%
    arrange(genes_id)  %>%
    dplyr::slice(1) %>%
    
    # reorder factor levels for appearance in plot legend
    mutate(genes_id = factor(genes_id, levels = c('1to5kb', 'promoter', '5UTR', 'exon', 'intron',
                                                  'intronexonboundary', '3UTR', 'intergenic')),
           group = 'Transcripts') %>%
    
    #plot
    ggplot(aes(x = cpg, y = genes_tx_id, fill = genes_id)) +
    geom_tile(color = '#969696', alpha = 0.6) +
    facet_grid(group~.) +
    guides(fill=guide_legend(ncol=2, override.aes=list(alpha=0.7, color = NA))) +
    theme_bw(base_size = font_size) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), 
          plot.margin = margin(l=-0.1,unit="cm"),
          panel.grid = element_blank()) +
    scale_fill_brewer(na.value = '#f7f7f7', breaks = c('1to5kb', 'promoter', '5UTR', 'exon', 
                                                        'intron', 'intronexonboundary', '3UTR',
                                                        'intergenic'),
                      palette = 'Paired', direction = -1) +
    scale_y_discrete(expand = c(0,0)) +
    scale_x_discrete(expand = c(0,0)) +
    labs(x = '', y = '', fill = '')
  
  # the size of each plot depends on the number of unique transcripts
  # More transcripts means a longer transcript track
  # so the plot height depends on the number of transcripts
  
  # number of transcripts
  n_trans <- betas_gene$genes_tx_id %>% str_split(', ') %>% unlist() %>% unique %>% length()
  
  # size of p1
  p1_h <- 35
  
  # combine tracks plot
  ggarrange(p1, p2, p3, ncol = 1, heights = c(p1_h-(n_trans-7), 7, n_trans))
  
}


```

##cpg
```{r}
plot_cpgs <- function(cpg_name, font_size = 12, point_size = 1) {
  # gene is the gene symbol to match plot
  # betas is the methylation data
  # pDat is the sample information which includes important group info to plot by
  # annotation is links cpgs to genes and other cpg annotation to plot

  # filter to gene in annotation
  
  anno_gene <- anno %>% filter(cpg %in% cpg_name)
  
  anno_gene <- anno_gene %>% filter(cpg %in% rownames(betas)) %>% 
    arrange(start)
    # generate mean and standard deviation of each cell type / trimester
  if (length(cpg_name) ==1)
  {
    betas_gene <- t(betas[rownames(betas) %in% cpg_name ,]*100)
    betas_gene <- t(betas_gene) %>% as_tibble
    colnames(betas_gene) <- cpg_name
    betas_gene <- betas_gene %>% mutate(Sample_Name = colnames(betas))%>%
    
    # reshape into longer format
    pivot_longer(cols = -Sample_Name,
                 names_to = 'cpg', 
                 values_to = 'beta')  %>%
    
    # add tissue and trimester info
    left_join(pDat %>% select(Sample_Name, Trimester, Tissue), by = 'Sample_Name') %>%
    
    # calculate mean and sd for each cpg for each group
    group_by(Tissue, Trimester, cpg) %>%
    summarize(mean = mean(beta),
              sd = sd(beta)) %>%
    mutate(lower = mean-sd, upper = mean+sd) %>%
    
    # add cpg info
    left_join(anno_gene, by = 'cpg') %>%
    
    # order on cpg position
    ungroup() %>%
    arrange(desc(start)) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           Trimester_Tissue = paste0(Trimester, ' - ', Tissue))
    }
    
  else {betas_gene <- t(betas[anno_gene$cpg,]*100) %>% as_tibble %>% 
    mutate(Sample_Name = colnames(betas)) %>%
    
    # reshape into longer format
    pivot_longer(cols = -Sample_Name,
                 names_to = 'cpg', 
                 values_to = 'beta')  %>%
    
    # add tissue and trimester info
    left_join(pDat %>% select(Sample_Name, Trimester, Tissue), by = 'Sample_Name') %>%
    
    # calculate mean and sd for each cpg for each group
    group_by(Tissue, Trimester, cpg) %>%
    summarize(mean = mean(beta),
              sd = sd(beta)) %>%
    mutate(lower = mean-sd, upper = mean+sd) %>%
    
    # add cpg info
    left_join(anno_gene, by = 'cpg') %>%
    
    # order on cpg position
    ungroup() %>%
    arrange(desc(start)) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           Trimester_Tissue = paste0(Trimester, ' - ', Tissue))
    }

     p1 <- betas_gene %>%
    ggplot() +
    
    # geom_ribbon is to generate the alternating shaded background
    geom_ribbon(aes(x = as.numeric(cpg), ymin = 0, ymax = 25),
                fill = 'grey', alpha = 0.25)+
    geom_ribbon(aes(x = as.numeric(cpg), ymin = 50, ymax = 75),
                fill = 'grey', alpha = 0.25)+
    
    # geom_linerange and geom_point is for the actual methylation data
    geom_linerange(alpha = 0.5, size = point_size, 
                   aes(x = cpg, ymin =lower, ymax = upper, color = Tissue)) +
    geom_point(alpha = 0.75, aes(x = cpg, y = mean, color = Tissue), size = point_size*2) +
    
    # we want to facet by trimester
    facet_grid(rows = vars(Trimester)) +
    
    # cosmetics
    theme_bw(base_size = font_size) +
    theme(axis.text.x = element_blank(), 
          axis.ticks.x = element_blank(),
          panel.grid.minor.y = element_blank(),
          panel.grid.major.y = element_blank(),
          plot.margin=margin(l=-0.1,unit="cm"),
          axis.title.y = element_text(angle = 0, vjust = 0.5)) +
    scale_y_continuous(limits = c(0, 100), expand = c(0, 0)) +
    scale_x_discrete(expand = c(0.01,0.01))+
    scale_color_manual(values= color_code_tissue[unique(betas_gene$Tissue)]) +
    labs(y = '% methylation', x = '', color = '', title = '') 
    
    p2 <- betas_gene %>% 
    
    # First we need to process the annotation data
    mutate(# absent/presence for different genomic elements
           enhancer = !is.na(enhancers_id),
           pmd = !is.na(pmd_id),
           imprinted_gene_placenta = !is.na(imprinted_gene_placenta),
           imprinted_gene_general = !is.na(imprinted_gene_general),
           imprinted_dmr_general = !is.na(imprinted_dmr_general),
           imprinted_dmr_placenta = !is.na(imprinted_dmr_placenta)) %>%
    mutate_if(is.logical, as.character) %>%
    select(cpg, enhancer, pmd, imprinted_gene_general:imprinted_dmr_placenta, cpg_id, start) %>%
    
    # reshape
    pivot_longer(cols = c(-cpg, -start),
               names_to = 'cpg_element', 
               values_to = 'presence') %>%
    distinct() %>%
    
    # arrange plot orde for each track
    arrange(desc(start)) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           cpg_element = factor(cpg_element, 
                                levels = c('enhancer', 'pmd', 'imprinted_gene_placenta',
                                           'imprinted_gene_general', 'imprinted_dmr_general',
                                           'imprinted_dmr_placenta', 'cpg_id')),
           presence = factor(ifelse(presence == "TRUE", 'Present',
                                    ifelse(presence == "FALSE", 'Absent', presence)),
                             levels = c('Present', 'Absent', ' ', '  ', 
                                        'sea', 'shelf', 'shore', 'island')),
           # for facet label
           group = 'Annotations') %>%
    
    # plot code
    ggplot(aes(x = cpg, y = cpg_element, fill = presence)) +
    scale_fill_manual(values = c('Present' = '#cccccc', 'Absent' = '#f7f7f7', 
                                 ' ' = 'white', '  ' = 'white',
                                 'sea' = '#ffffcc', 'shelf' = '#a1dab4', 'shore' = '#41b6c4',
                                 'island' = '#225ea8'), drop = T) +
    geom_tile(color = '#969696') +
    guides(fill=guide_legend(ncol=3, override.aes=list(colour=NA))) +
    facet_grid(group~.) +
    theme_bw(base_size = font_size) +
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          legend.key = element_rect(colour = NA), 
          plot.margin = margin(l=-0.1,unit="cm")) +
    scale_y_discrete(expand = c(0,0)) +
    scale_x_discrete(expand = c(0,0)) +
    labs(x = '', y = '', fill = '')
  
  # transcript annotations
  
  # cpgs can have more than one mapping to a gene element per transcript
  # cpgs mapped to intronexonboundaries can also be part of exons and introns
  # 5 UTR and 3 UTR can also overlap with exons and introns
  
  # If there are multiple mappings then only the following will be displayed in order of priority:
  # 5UTR / 3UTR, exon / intron, intronexonboundary
  
  p3 <- betas_gene %>% 
    select(cpg, end, genes_id, genes_tx_id) %>%
    distinct() %>%
    
    # split comma separate list and put each entry into a new row
    mutate(genes_tx_id = str_split(genes_tx_id, ', '),
           genes_id = str_split(genes_id, ', ')) %>%
    unnest(c(genes_tx_id, genes_id)) %>%
    
    # order transcripts by position of start
    arrange(end) %>%
    mutate(genes_tx_id = factor(genes_tx_id, levels = unique(genes_tx_id))) %>%
    
    # fill all combinations of cpg and gene_tx_id with NA if missing, so that they show in plot
    complete(cpg, genes_tx_id) %>%
    
    # filter out to unique mappings using the priority described above
    # strategy is to order by factor level and take the first occurence
    mutate(genes_id = factor(genes_id, levels = c('1to5kb', 'promoter', '5UTR', '3UTR', 'exon', 
                                                  'intron', 'intronexonboundary', 'intergenic'))) %>%
    
    group_by(cpg, genes_tx_id) %>%
    arrange(genes_id)  %>%
    dplyr::slice(1) %>%
    
    # reorder factor levels for appearance in plot legend
    mutate(genes_id = factor(genes_id, levels = c('1to5kb', 'promoter', '5UTR', 'exon', 'intron',
                                                  'intronexonboundary', '3UTR', 'intergenic')),
           group = 'Transcripts') %>%
    
    #plot
    ggplot(aes(x = cpg, y = genes_tx_id, fill = genes_id)) +
    geom_tile(color = '#969696', alpha = 0.6) +
    facet_grid(group~.) +
    guides(fill=guide_legend(ncol=2, override.aes=list(alpha=0.7, color = NA))) +
    theme_bw(base_size = font_size) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), 
          plot.margin = margin(l=-0.1,unit="cm"),
          panel.grid = element_blank()) +
    scale_fill_brewer(na.value = '#f7f7f7', breaks = c('1to5kb', 'promoter', '5UTR', 'exon', 
                                                        'intron', 'intronexonboundary', '3UTR',
                                                        'intergenic'),
                      palette = 'Paired', direction = -1) +
    scale_y_discrete(expand = c(0,0)) +
    scale_x_discrete(expand = c(0,0)) +
    labs(x = '', y = '', fill = '')
  
  # the size of each plot depends on the number of unique transcripts
  # More transcripts means a longer transcript track
  # so the plot height depends on the number of transcripts
  
  # number of transcripts
  n_trans <- betas_gene$genes_tx_id %>% str_split(', ') %>% unlist() %>% unique %>% length()
  
  # size of p1
  p1_h <- 35
  
  # combine tracks plot
  ggarrange(p1, p2, p3, ncol = 1, heights = c(p1_h-(n_trans-7), 7, n_trans))
}
```

