---
title: "shiny_DMR"
output: html_document
---

#set up
```{r}
#install.packages('rsconnect')
#install.packages('shinyjs')
library(rsconnect)
library(shiny)
rsconnect::setAccountInfo(name='yifan7', token='52E1037550A7AA68252579A98EE8AA38', secret='+NlrJysSVZHa5LQQb9slmYRDXExPfjF0QbxMjUH1')
library(DT)
library(shinyjs)
library(shinythemes)
```

#UI
```{r}
ui <- fluidPage(
  theme = shinythemes::shinytheme("lumen"),
  
  useShinyjs(),
  #main title
  titlePanel("Placental Cell Methylome Browser"),
  hr(), #create a line
  
  #UI layout
  sidebarLayout(
      sidebarPanel(
          
           # relative width out of 12 units
           width = 3,
       
           #app description
           p('Welcome to the Placental Cell Methylome Browser!'),
           p('To start, select CpGs or a gene.', span(tags$a(href = '#footer_1', 
                                                             tags$sup(1), 
                                                             style = 'color:blue'))),
           hr(),
           
           #tabset panel
           tabsetPanel(type = 'tabs', id = 'tabs',
                       
           #page of gene input 
           tabPanel('Gene Input',
                    #gene search list
                    selectizeInput('gene_name', 'Gene', 
                          choices = NULL, 
                          multiple = FALSE,
                          options = list(placeholder='Enter a gene name',
                                         onInitialize = I('function() { this.setValue(""); }'))),
                    #action button to submit and reset input
                    actionButton("submit_gene_input", "Submit"),
                    actionButton('reset_gene_input','Reset'),
                    br(), # add a space line
                    hr(), # create a horizontal line
           
                    p('If a specific gene is selected, then you can also display specific associated-CpGs'),
           
                    #cpgs search list related to selected gene
                    selectizeInput('cpg', 'CpGs related to the selected gene', 
                          choices = NULL,
                          multiple = TRUE),
                    #action button to submit and reset input
                    actionButton('submit_cpg_gene_input', 'Submit'),
                    actionButton('reset_cpg_gene_input','Reset'),
                    br(), 
                    hr(), 
                    #numeric inputs to control the number of cpgs to show
                    numericInput('cpg_number_input', 'CpG number to plot', value = 50),
                    actionButton('submit_cpg_number','Submit'),
                    #numeric inputs to control which sets of cpgs to show
                    numericInput('first_cpg', 'First CpG to plot', value = 1),
                    numericInput('end_cpg', 'End CpG to plot', value = 50),
                    actionButton('submit_cpg_set','Submit')
                    ),
                       
            #page of cpgs input
            tabPanel('CpG Input',
                    #cpg search list
                    selectizeInput('cpg_name', 'CpG Sites', 
                          choices = NULL,
                          multiple = TRUE,
                          options = list(placeholder='Enter CpG site names' )),
                    #action button to submit and reset input
                    actionButton("submit_cpg_input", "Submit"),
                    actionButton('reset_cpg_input','Reset')),
                       
            #page of region input
                    tabPanel('Region Input',
                    #chromosome 
                    selectizeInput('chr','Chromosome',
                          choices = NULL,
                          multiple = FALSE,
                          options = list(placeholder = 'Enter a chromosome')),
                    #start point
                    numericInput('start', 'Start', value = NULL),
                    #end point
                    numericInput('end', 'End', value = NULL),
                    
                    #action button to submit and reset input
                    actionButton("submit_region_input", "Submit"),
                    actionButton('reset_region_input','Reset')
                                )
                       ),
           
           #create a blank line
           br(),
           #create a horizontal line
           hr(),
           #checkbox to select information shown in datatable
           checkboxGroupInput('checkbox', 
                              'Columns to show in CpG annotation table', 
                              names(anno), 
                              selected = names(anno)[-10:-12])
           ), 
  
  mainPanel(
    hr(),

    textOutput('text'),

    h6("CpG Information. Click on a CpG ID to show a boxplot at the bottom."),
  

    #create a data table to show cpg site information 
    div(dataTableOutput('dt_anno_cpg'), style = "font-size:85%"),
    
    hr(),
    
    # GENE PLOT
    plotOutput('annoplot', height = 900),
    
    hr(),
    
    # INDIVIDUAL CPG PLOT
    plotOutput('boxplot_selected_cpg')
  
    #boxplot that shows tissue and trimester
    #selectInput('num','Number of CpGs to plot', choices = c(25, 50, 75,100)),
    #plotOutput('boxplot_total',  height = '650px'),
    )
  ),
  
  #footer
  hr(),

  tags$h6(paste("[1] Only associated CpGs will be displayed when selecting a gene.",
                "This CpG-gene mapping is based on the Illumina-provided EPIC array annotation.", 
                id = "#footer_1")),

  tags$h6(paste("[1] When selecting a gene, all CpGs that are mapped to any component of an associated transcript will be shown. Transcript components include 1-5Kb upstream of the TSS, the promoter (< 1Kb upstream of the TSS), 5’UTR, exons, introns, and 3’UTRs. Everything else are intergenic regions")),
  
  tags$h6(paste("[2] All DNA methylation data here is 850k array data. See primary publication for details on processing.")),
  
  tags$h6(paste("[3] Annotations are in hg19 coordinates.")),
  
  tags$h6(paste("[4] Differential methylation is defined as a bonferroni-adjusted p-value < 0.01, and an absolute difference in mean DNAm > 25%.")),
  
  tags$h6(paste("[5] PMD regions coordinates are taken from D. I. Schroeder et al. 2013."))
  

)


```

#Server
```{r}
server <- function(input, output, session) {
    ############  searching lists ############
    #searching list of cpgs
    updateSelectizeInput(session = session, inputId = 'cpg_name', 
                         choices = cpg$value, server = TRUE)
    #searching list of genes
    updateSelectizeInput(session = session, inputId = 'gene_name', 
                         choices = gene$gene, selected = '',server = TRUE)
    #searching list of chromosomes
    chr <- unique(anno$chr) %>% str_sort(numeric = TRUE) %>% as.tibble()
    updateSelectizeInput(session = session, inputId = 'chr',
                         choices = chr$value, selected = '', server = TRUE)
    
    ############  reset button ############  
    #reset the input and outputs when click the reset button
    observeEvent(input$reset_cpg_input, {reset('cpg_name')})
    observeEvent(input$reset_gene_input, {reset('gene_name') & reset('cpg_number_input') & reset('first_cpg') & reset('end_cpg')})
    observeEvent(input$reset_region_input, {reset('chr') & reset('start') & reset('end')})
    observeEvent(input$reset_gene_input, {input_value$input_value <- FALSE})
    observeEvent(input$reset_cpg_input, {input_value$input_value <- FALSE})
    observeEvent(input$reset_region_input, {input_value$input_value <- FALSE})
    #reset the input and output when switch between tabs
    observeEvent(input$tabs, {input_value$input_value <- FALSE})
    observeEvent(input$tabs, reset('cpg_name') & reset('gene_name') & reset('chr') & reset('start') & reset('end') & reset('cpg_number_input') & reset('first_cpg') & reset('end_cpg'))
    
    ############  submit button ############
    #store inputs into reactive values
    input_value <- reactiveValues(input_value = FALSE)
    #get input values when click the submit button
    observeEvent(input$submit_cpg_input, {input_value$input_value <- input$submit_cpg_input})
    observeEvent(input$submit_gene_input, {input_value$input_value <- input$submit_gene_input})
    observeEvent(input$submit_gene_input, {reset('cpg_number_input')})
    observeEvent(input$submit_region_input, {input_value$input_value <- input$submit_region_input})
  
    ############  datatable output ############
    #make an empty datatable as default
    empty_df <- anno[0,] %>% rename_columns() %>% select(CpG:`Relation to Transcript Features`)
    #datatable to show cpg annotation information 
    output$dt_anno_cpg <- DT::renderDataTable({
      #output an empty datatable when no inputs
      if (input_value$input_value == FALSE)
        return(DT:: datatable(empty_df, 
                              options = list(paging = FALSE, 
                                             searching = FALSE)))
      #output a datatable when input a gene name
      isolate(if(input$tabs == 'Gene Input')
      {DT::datatable(make_datatable(gene_name = input$gene_name, cpg_name = NULL,Chr = NULL, Start = NULL, End = NULL)[,input$checkbox, 
                                                                drop = FALSE], 
                    options = list(pageLength = 5),
                    selection = 'single',
                    rownames = FALSE) %>% 
        formatStyle(1, cursor = 'pointer', color = 'blue')}
      #output a datatable when input cpgs
      else if(input$tabs == 'CpG Input')
        {DT::datatable(make_datatable(gene_name = NULL, cpg_name = input$cpg_name, Chr = NULL, Start = NULL, End = NULL)[,input$checkbox, 
                                                                drop = FALSE], 
                    options = list(pageLength = 5),
                    selection = 'single',
                    rownames = FALSE) %>% 
        formatStyle(1, cursor = 'pointer', color = 'blue')}
      #output a datatable when input a region
      else if(input$tabs == 'Region Input')
        {DT::datatable(make_datatable(gene_name = NULL, cpg_name = NULL, Chr = input$chr, Start = input$start, End = input$end)[,input$checkbox, 
                                                                drop = FALSE], 
                    options = list(pageLength = 5),
                    selection = 'single',
                    rownames = FALSE) %>% 
        formatStyle(1, cursor = 'pointer', color = 'blue')
        
        })
      }
    )
    ############  annoplot output: default to show first 50 cpgs############
    #gene cpg plots
    output$annoplot <- renderPlot(
      {
      #return empty when no inputs
      if (input_value$input_value  == FALSE) return()
      isolate(
      #output a plot when input a gene
      if(input$tabs == 'Gene Input')
        {
        annoPlot_with_tracks(cpg_name = NULL, gene_name = input$gene_name, Chr = NULL, Start = NULL, End = NULL)
      }
      #output a plot when input cpgs
      else if(input$tabs == 'CpG Input')
        {
        annoPlot_with_tracks(cpg_name = input$cpg_name, gene_name = NULL, Chr = NULL, Start = NULL, End = NULL)
      }
      #output a plot when input a region
      else if(input$tabs == 'Region Input')
        {
        annoPlot_with_tracks(cpg_name =  NULL, gene_name = NULL, Start = input$start, End = input$end, Chr = input$chr)
        })
      
          
      }

    )
    ############  annoplot output: choose cpg number ############
    #get the input when click submit button
    get_gene_input <- eventReactive(input$submit_gene_input, {input$gene_name})
    get_cpg_input <- eventReactive(input$submit_cpg_input, {input$cpg_name})
    get_region_chr<- eventReactive(input$submit_region_input, {input$chr})
    get_region_start<- eventReactive(input$submit_region_input, {input$start})
    get_region_end<- eventReactive(input$submit_region_input, {input$end})
    #store cpg number into reactive value
    cpg_number <- reactiveValues(cpg_number = 50)
    #get the cpg number input
    observeEvent(input$submit_cpg_number, {cpg_number$cpg_number <- input$cpg_number_input})
    #reset cpg number into default(50) when select another gene
    observeEvent(input$submit_gene_input, {cpg_number$cpg_number = 50})
    #update the plot when change the cpg number to show
    observeEvent(input$submit_cpg_number, {output$annoplot <- renderPlot(
      if (input_value$input_value  == FALSE) return()
      else if(input$tabs == 'Gene Input')
        {
        annoPlot_with_tracks(cpg_name = NULL, gene_name = get_gene_input(), Chr = NULL, Start = NULL, End = NULL, cpg_number = cpg_number$cpg_number)
        
      }
      #output a plot when input cpgs
      else if(input$tabs == 'CpG Input')
        {
        annoPlot_with_tracks(cpg_name = get_cpg_input(), gene_name = NULL, Chr = NULL, Start = NULL, End = NULL)
      }
      #output a plot when input a region
      else if(input$tabs == 'Region Input')
        {
        annoPlot_with_tracks(cpg_name =  NULL, gene_name = NULL, Start = get_region_start(), End = get_region_end(), Chr = get_region_chr())
        }
      )
    })
    
    ############  annoplot output: choose a cpg set############
    #store cpg_start and cpg_end into reactive value
    cpg_start <- reactiveValues(cpg_start = 1)
    cpg_end <- reactiveValues(cpg_start = 50)
    #get the cpg set input
    observeEvent(input$submit_cpg_set, {cpg_start$cpg_start <- input$first_cpg
                                        cpg_end$cpg_end <- input$end_cpg})
    #update the plot when change the cpg number to show
    observeEvent(input$submit_cpg_set, {output$annoplot <- renderPlot(
      if (input_value$input_value  == FALSE) return()
      else if(input$tabs == 'Gene Input')
        {
        annoPlot_with_tracks(cpg_name = NULL, gene_name = get_gene_input(), Chr = NULL, Start = NULL, End = NULL, first_cpg = cpg_start$cpg_start, end_cpg = cpg_end$cpg_end )
        
      }
      #output a plot when input cpgs
      else if(input$tabs == 'CpG Input')
        {
        annoPlot_with_tracks(cpg_name = get_cpg_input(), gene_name = NULL, Chr = NULL, Start = NULL, End = NULL)
      }
      #output a plot when input a region
      else if(input$tabs == 'Region Input')
        {
        annoPlot_with_tracks(cpg_name =  NULL, gene_name = NULL, Start = get_region_start(), End = get_region_end(), Chr = get_region_chr())
        }
      )
    })
    
    ############  boxplot output ############
    #get cpg sites related to the selected gene
    cpg_related_to_gene <- reactive(find_cpg_ids(get_gene_input()))
    #searching list (cpg sites related to selected gene)
    observe(
      {
        updateSelectizeInput(session = session, 
                             inputId = 'cpg', 
                             choices = cpg_related_to_gene(), 
                             server = TRUE, 
                             selected = '')
        
      })
    
    ##retrive cpg name when click the datatable
    get_cpg_from_dt <- eventReactive(input$submit_cpg_gene_input, {input$cpg})
    
    #produce a boxplot based on the selected cpg from datatable
    observeEvent(input$dt_anno_cpg_cell_clicked, 
                 {
                   c_info = input$dt_anno_cpg_cell_clicked
                   
                   output$boxplot_selected_cpg <- renderPlot(
      {
        if (!is.null(c_info$value))
        {
          boxplot_selected(c_info$value)
        }
        else if(!is.null(get_cpg_from_dt()))
        {
          boxplot_selected(get_cpg_from_dt())
        }
      }
    )
    })
    #produce a boxplot based on the selected cpg using search box
    observeEvent(input$submit_cpg_gene_input, {output$boxplot_selected_cpg <- renderPlot(
      {
        boxplot_selected(get_cpg_from_dt())
      }
    )})
    
    #reset the associated cpg inputs in search box
    observeEvent(input$reset_cpg_gene_input, {reset('cpg')})
    #reset the boxplot when click the reset button
    observeEvent(input$reset_cpg_gene_input, {output$boxplot_selected_cpg <- renderPlot(
      {
        return()
      }
    )})
    #reset the boxplot when switch between tabs
    observeEvent(input$tabs, {output$boxplot_selected_cpg <- renderPlot(
      {
        return()
      }
    )})
    #reset the boxplot when change into another gene
    observeEvent(input$submit_gene_input, {output$boxplot_selected_cpg <- renderPlot(
      {
        return()
      }
    )})
    #reset the boxplot when reset gene input
    observeEvent(input$reset_gene_input, {output$boxplot_selected_cpg <- renderPlot(
      {
        return()
      }
    )})
   
    
}

```

#Run App
```{r}
shinyApp(ui, server)
```

