---
title: "shiny_DMR"
output: html_document
---

#set up
```{r}
#install.packages('rsconnect')
#install.packages('shinyjs')
library(rsconnect)
library(shiny)
rsconnect::setAccountInfo(name='yifan7', token='52E1037550A7AA68252579A98EE8AA38', secret='+NlrJysSVZHa5LQQb9slmYRDXExPfjF0QbxMjUH1')
library(DT)
library(shinyjs)
library(shinythemes)
```

#UI

```{r}
ui <- fluidPage(
  theme = shinythemes::shinytheme("lumen"),
  
  useShinyjs(),
  #main title
  titlePanel("Welcome to the Placental Cell Methylome Browser!"),
  p('Scroll down to see detailed instructions.'),
  
  #UI layout
  sidebarLayout(
    sidebarPanel(
    
      # relative width out of 12 units
      width = 3,
      
      tags$h4(
        tags$strong('Plot input:')),
      
      p('Select a gene, specific CpGs, or a genomic region:', 
        span(tags$a(href = '#footer_1', tags$sup(1), style = 'color:blue'))),
      
      tabsetPanel(type = 'tabs', id = 'tabs',
                       
      #page of gene input 
      tabPanel('Gene Input',
               
               #gene search list
               selectizeInput(
                 'gene_name', '',
                 choices = NULL,
                 multiple = FALSE,
                 options = list(placeholder='Enter a gene name',
                                onInitialize = I('function() { this.setValue(""); }'))),
               #action button to submit and reset input
               actionButton("submit_gene_input", "Plot"),
               actionButton('reset_gene_input','Clear')),
      
      #page of cpgs input
      #tabPanel('CpG Input',
               
               #cpg search list
      #         selectizeInput('cpg_name', '',
      #                        choices = NULL,
      #                        multiple = TRUE,
      #                        options = list(placeholder='Enter CpG site names' )),
               
               #action button to submit and reset input
      #         actionButton("submit_cpg_input", "Plot"),
      #         actionButton('reset_cpg_input','Clear')),
      
      #page of region input
      tabPanel('Region Input',
               
               #chromosome 
               selectizeInput('chr','Chromosome',
                              choices = NULL,
                              multiple = FALSE,
                              options = list(placeholder = 'Enter a chromosome')),
               
               #start point
               numericInput('start', 'Start', value = NULL),
               #end point
               numericInput('end', 'End', value = NULL),
               
               #action button to submit and reset input
               actionButton("submit_region_input", "Plot"),
               actionButton('reset_region_input','Clear'))),
      hr(),
      
      tags$h4(
        tags$strong('Customize plot')),
      radioButtons('checkbox_trimester',
                   'Include samples (trimester)',
                   setNames(c('First', 'Third'), c('First Trimester', 'Third Trimester (Term)')),
                   selected = 'Third'),
      checkboxGroupInput('checkbox_tissue',
                         'Include samples (cell type):',
                         setNames(unique(pDat$Tissue), gsub(' cs', '', unique(pDat$Tissue))),
                         selected = unique(pDat$Tissue)),
      
      
      checkboxGroupInput('checkbox_sex',
                         'Include samples (sex):',
                         setNames(c('F', 'M'), c('Female', 'Male')),
                         selected = unique(pDat$Sex)),
      
      
      
      sliderInput('range_cpg', 
                  'Select number of CpGs to plot (only applies to Condensed View):',
                  min = 1, max = 500, value = c(500)),
      actionButton('submit_cpg_set','Submit'),
      br(),
      hr(),
      
      tags$h4(
        tags$strong('Customize CpG annotation table')),
      
      checkboxGroupInput('checkbox',
                         'Show columns:',
                         
                         choices = names(rename_columns(anno))[-13:-20],
                         inline = FALSE,
                         selected = names(rename_columns(anno))[-13:-20])),
    
    mainPanel(
      tabsetPanel(
        type = 'tabs',
      
        # First tab
        tabPanel("Condensed View",
                 
                 fluidRow(
                   br(),
                   tags$h4(
                     tags$strong(
                       'Adjust heights of individual data tracks:'
                       )
                     ),
                   column(
                     12,
                     wellPanel(
                       fluidRow(
                       column(
                         3, 
                         sliderInput(
                           'condensed_plot_height_1',
                           'DNA methylation',
                           min = 0, max = 25, value = c(10), step = 1)
                         ),
                       column(
                         3,
                         sliderInput(
                           'condensed_plot_height_2',
                           'Difference from CpG mean',
                           min = 0, max = 25, value = c(5), step = 1)),
                       column(
                         3, 
                         sliderInput(
                           'condensed_plot_height_3',
                           'Annotations',
                           min = 0, max = 25, value = c(5), step = 1)),
                       column(
                         3, 
                         sliderInput(
                           'condensed_plot_height_4',
                           'Transcripts',
                           min = 0, max = 25, value = c(5), step = 1)
                       )
                     )
                   ))
                   ),
                 
                 br(),
                 textOutput('text'),
                 plotOutput('annoplot', height = 1000)),
        
        tabPanel("Expanded View",
                 
                 fluidRow(
                   br(),
                   tags$h4(
                     tags$strong(
                       'Adjust heights of individual data tracks:'
                       )
                     ),
                   column(
                     6, 
                     wellPanel(
                       fluidRow(
                         column(
                           6,
                           sliderInput(
                             'expanded_plot_height_1', 'Methylation',
                             min = 1, max = 10, value = c(3), step = 1),
                         ),
                         column(
                           6,
                           sliderInput(
                             'expanded_plot_height_2', 'Transcripts',
                             min = 1, max = 10, value = c(1), step = 1)
                         )
                       )
                     )
                   )
                   ),
                 
                 
                 br(),
                 plotOutput('expanded', height = 1000)),
        
        tabPanel("CpG Boxplots",
                 p('Show boxplots for specific CpGs:'),
                 
                 #cpgs search list related to the input
                 selectizeInput('cpg', 'CpGs related to the input',
                                choices = NULL,
                                multiple = TRUE),
                 
                 #action button to submit and reset input
                 actionButton('submit_cpg_gene_input', 'Submit'),
                 actionButton('reset_cpg_gene_input','Reset'),
                         
                 plotOutput('boxplot_selected_cpg')
                         
                         )))),
  
  tags$h3('Annotation Table'),
  fluidRow(
    
    
    
    div(
      dataTableOutput('dt_anno_cpg', width = "98%"), 
      style = "font-size:85%",
      align = 'center')),
  
  #footer
  hr(),

  HTML(paste("[1] Only associated CpGs will be displayed when selecting a gene.This CpG-gene mapping is based on the Illumina-provided EPIC array annotation.",
             "[2] When selecting a gene, all CpGs that are mapped to any component of an associated transcript will be shown. Transcript components include 1-5Kb upstream of the TSS, the promoter (< 1Kb upstream of the TSS), 5’UTR, exons, introns, and 3’UTRs. Everything else are intergenic regions", 
             "[3] All DNA methylation data here is 850k array data. See primary publication for details on processing.",
             "[4] Annotations are in hg19 coordinates.",
             "[5] Differential methylation is defined as a bonferroni-adjusted p-value < 0.01, and an absolute difference in mean DNAm > 25%.",
             "[6] PMD regions coordinates are taken from D. I. Schroeder et al. 2013.",
             "[7] Differential methylation by cell type was conducted separated in first and third trimester samples for autosomal CpGs. Differentially methylated cytosines (DMCs) are defined as a bonferroni adjusted p-value < 0.01 and a mean difference in methylation > 10%. Differential methylation was conducted with a one-versus-all design.",
             
             
             sep = " <br> "))
)
```

#Server

```{r}
server <- function(input, output, session) {
    ############  searching lists ############
    #searching list of cpgs
    updateSelectizeInput(session = session, inputId = 'cpg_name', 
                         choices = cpg$value, server = TRUE)
    #searching list of genes
    updateSelectizeInput(session = session, inputId = 'gene_name', 
                         choices = gene$gene, selected = '',server = TRUE)
    #searching list of chromosomes
    chr <- unique(anno$chr) %>% str_sort(numeric = TRUE) %>% as.tibble()
    updateSelectizeInput(session = session, inputId = 'chr',
                         choices = chr$value, selected = '', server = TRUE)
    
    #get cpg sites related to the selected gene
    cpg_related_to_gene <- reactive(find_cpg_ids(get_gene_input()))
    cpg_related_to_region <- reactive(find_cpg_from_region(Chr = get_region_chr(), Start = get_region_start(), End = get_region_end()))
    #searching list (cpg sites related to selected gene)
    observe(
      { if(input$tabs == 'Gene Input')
        {updateSelectizeInput(session = session, 
                             inputId = 'cpg', 
                             choices = cpg_related_to_gene(), 
                             server = TRUE, 
                             selected = '')}
        else if(input$tabs == 'Region Input')
        {updateSelectizeInput(session = session, 
                             inputId = 'cpg', 
                             choices = cpg_related_to_region(), 
                             server = TRUE, 
                             selected = '')}
        else if(input$tabs == 'CpG Input')
        {updateSelectizeInput(session = session, 
                             inputId = 'cpg', 
                             choices = get_cpg_input(), 
                             server = TRUE, 
                             selected = '')}
        
        
      })
    
    ############  reset button ############  
    #reset the input and outputs when click the reset button
    observeEvent(input$reset_cpg_input, {reset('cpg_name')})
    observeEvent(input$reset_gene_input, {
      reset('gene_name') & reset('cpg_number_input') & reset('range_cpg') & reset('cpg')})
    observeEvent(input$reset_region_input, {
      reset('chr') & reset('start') & reset('end')& reset('cpg_number_input_region')})
    observeEvent(input$reset_gene_input, {input_value$input_value <- FALSE})
    observeEvent(input$reset_cpg_input, {input_value$input_value <- FALSE})
    observeEvent(input$reset_region_input, {input_value$input_value <- FALSE})
    #reset the input and output when switch between tabs
    observeEvent(input$tabs, {input_value$input_value <- FALSE})
    observeEvent(input$tabs, 
                 reset('cpg_name') & reset('gene_name') & reset('chr') & reset('start') & 
                   reset('end') & reset('cpg_number_input') & reset('range_cpg') & reset('cpg'))
    
    ############  submit button ############
    #store inputs into reactive values
    input_value <- reactiveValues(input_value = FALSE)
    #get input values when click the submit button
    observeEvent(input$submit_cpg_input, {input_value$input_value <- input$submit_cpg_input})
    observeEvent(input$submit_gene_input, {input_value$input_value <- input$submit_gene_input})
    observeEvent(input$submit_gene_input, {
      reset('cpg_number_input')& reset('range_cpg') &reset('cpg')})
    observeEvent(input$submit_region_input, {input_value$input_value <- input$submit_region_input})
    observeEvent(input$submit_region_input, {reset('cpg_number_input_region') & reset('cpg')})
    
    
    ############  datatable output ############
    #make an empty datatable as default
    empty_df <- anno[0,] %>% rename_columns() %>% select(CpG:`Relation to Transcript Features`)
    #datatable to show cpg annotation information 
    
    output$dt_anno_cpg <- DT::renderDataTable({
      #output an empty datatable when no inputs
      if (input_value$input_value == FALSE)
        return(DT:: datatable(empty_df, 
                              options = list(paging = FALSE, 
                                             searching = FALSE)))
      #output a datatable when input a gene name
      isolate(if(input$tabs == 'Gene Input') {
        
        DT::datatable(
          make_datatable(gene_name = input$gene_name, 
                         cpg_name = NULL,
                         Chr = NULL, Start = NULL, End = NULL)[,input$checkbox,drop = FALSE],
          options = list(pageLength = 5),
          selection = 'single',
          rownames = FALSE) %>%
          formatStyle(1, cursor = 'pointer', color = 'blue')}
        
        #output a datatable when input cpgs
        else if(input$tabs == 'CpG Input') {
          
          DT::datatable(
            make_datatable(gene_name = NULL, cpg_name = input$cpg_name, 
                           Chr = NULL, Start = NULL, End = NULL)[,input$checkbox, drop = FALSE], 
            options = list(pageLength = 5),
            selection = 'single',
            rownames = FALSE) %>%
            formatStyle(1, cursor = 'pointer', color = 'blue')
          } else if(input$tabs == 'Region Input') {#output a datatable when input a region
        
            DT::datatable(
              make_datatable(
                gene_name = NULL, cpg_name = NULL,
                Chr = input$chr, Start = input$start, End = input$end
                )[,input$checkbox, drop = FALSE],
              options = list(pageLength = 5),
              selection = 'single',
              rownames = FALSE
              ) %>%
              formatStyle(1, cursor = 'pointer', color = 'blue')
        
        })
      }
    )
    
    ############  annoplot output: default to show first 250 cpgs############
    #gene cpg plots
    output$annoplot <- renderPlot(
      {
      #return empty when no inputs
      if (input_value$input_value  == FALSE) return()
      #output a plot when input a gene
      else if(input$tabs == 'Gene Input')
        {
        annoPlot_with_tracks(cpg_name = NULL, gene_name = input$gene_name, 
                             Chr = NULL, Start = NULL, End = NULL, 
                             Tissue_type = input$checkbox_tissue,
                             Sex_type = input$checkbox_sex,
                             Trimester_type = input$checkbox_trimester,
                             plot_heights = c(input$condensed_plot_height_1,
                                              input$condensed_plot_height_2,
                                              input$condensed_plot_height_3,
                                              input$condensed_plot_height_4))
      }
      #output a plot when input cpgs
      else if(input$tabs == 'CpG Input')
        {
        annoPlot_with_tracks(cpg_name = input$cpg_name, gene_name = NULL, 
                             Chr = NULL, Start = NULL, End = NULL, 
                             Tissue_type = input$checkbox_tissue,
                             Sex_type = input$checkbox_sex,
                             Trimester_type = input$checkbox_trimester,
                             plot_heights = c(input$condensed_plot_height_1,
                                              input$condensed_plot_height_2,
                                              input$condensed_plot_height_3,
                                              input$condensed_plot_height_4))
      }
      #output a plot when input a region
      else if(input$tabs == 'Region Input')
        {
        annoPlot_with_tracks(cpg_name =  NULL, gene_name = NULL, 
                             Start = input$start, End = input$end, Chr = input$chr, 
                             Tissue_type = input$checkbox_tissue,
                             Sex_type = input$checkbox_sex,
                             Trimester_type = input$checkbox_trimester,
                             plot_heights = c(input$condensed_plot_height_1,
                                              input$condensed_plot_height_2,
                                              input$condensed_plot_height_3,
                                              input$condensed_plot_height_4))
        }
      
          
      }
    )
    ############  annoplot output: choose cpg number (given a gene input) ############
    #get the input when click submit button
    get_gene_input <- eventReactive(input$submit_gene_input, {input$gene_name})
    get_cpg_input <- eventReactive(input$submit_cpg_input, {input$cpg_name})
    get_region_chr<- eventReactive(input$submit_region_input, {input$chr})
    get_region_start<- eventReactive(input$submit_region_input, {input$start})
    get_region_end<- eventReactive(input$submit_region_input, {input$end})
    
    #store cpg number into reactive value
    cpg_number <- reactiveValues(cpg_number = 250)
    #get the cpg number input
    observeEvent(input$submit_cpg_number, {cpg_number$cpg_number <- input$cpg_number_input})
    #reset cpg number into default(50) when select another gene
    observeEvent(input$submit_gene_input, {cpg_number$cpg_number = 250})
    #update the plot when change the cpg number to show
    observeEvent(input$submit_cpg_number, {output$annoplot <- renderPlot(
      if (input_value$input_value  == FALSE) return()
      else if(input$tabs == 'Gene Input')
        {
        annoPlot_with_tracks(cpg_name = NULL, gene_name = get_gene_input(), 
                             Chr = NULL, Start = NULL, End = NULL, 
                             cpg_number = cpg_number$cpg_number, 
                             Tissue_type = input$checkbox_tissue,
                             Sex_type = input$checkbox_sex,
                             Trimester_type = input$checkbox_trimester,
                             plot_heights = c(input$condensed_plot_height_1,
                                              input$condensed_plot_height_2,
                                              input$condensed_plot_height_3,
                                              input$condensed_plot_height_4))
        
      }
      #output a plot when input cpgs
      else if(input$tabs == 'CpG Input')
        {
        annoPlot_with_tracks(cpg_name = get_cpg_input(), gene_name = NULL, 
                             Chr = NULL, Start = NULL, End = NULL, 
                             Tissue_type = input$checkbox_tissue,
                             Sex_type = input$checkbox_sex,
                             Trimester_type = input$checkbox_trimester,
                             plot_heights = c(input$condensed_plot_height_1,
                                              input$condensed_plot_height_2,
                                              input$condensed_plot_height_3,
                                              input$condensed_plot_height_4))
      }
      #output a plot when input a region
      else if(input$tabs == 'Region Input')
        {
        annoPlot_with_tracks(cpg_name =  NULL, gene_name = NULL, 
                             Start = get_region_start(), 
                             End = get_region_end(), 
                             Chr = get_region_chr(), 
                             Tissue_type = input$checkbox_tissue,
                             Sex_type = input$checkbox_sex,
                             Trimester_type = input$checkbox_trimester,
                             plot_heights = c(input$condensed_plot_height_1,
                                              input$condensed_plot_height_2,
                                              input$condensed_plot_height_3,
                                              input$condensed_plot_height_4))
        }
      )
    })
    ############  annoplot output: choose cpg number (given a region input) ############
    #store cpg number into reactive value
    cpg_number_region <- reactiveValues(cpg_number = 250)
    #get the cpg number input
    observeEvent(input$submit_cpg_number_region, {cpg_number_region$cpg_number <-
      input$cpg_number_input_region})
    #reset cpg number into default(50) when select another gene
    observeEvent(input$submit_region_input, {cpg_number_region$cpg_number = 250})
    #update the plot when change the cpg number to show
    observeEvent(input$submit_cpg_number_region, {output$annoplot <- renderPlot(
      if (input_value$input_value  == FALSE) return()
      else if(input$tabs == 'Gene Input')
        {
        annoPlot_with_tracks(cpg_name = NULL, gene_name = get_gene_input(),
                             Chr = NULL, Start = NULL, End = NULL, 
                             Tissue_type = input$checkbox_tissue,
                             Sex_type = input$checkbox_sex,
                             Trimester_type = input$checkbox_trimester,
                             plot_heights = c(input$condensed_plot_height_1,
                                              input$condensed_plot_height_2,
                                              input$condensed_plot_height_3,
                                              input$condensed_plot_height_4))
        
      }
      #output a plot when input cpgs
      else if(input$tabs == 'CpG Input')
        {
        annoPlot_with_tracks(cpg_name = get_cpg_input(), gene_name = NULL, 
                             Chr = NULL, Start = NULL, End = NULL, 
                             Tissue_type = input$checkbox_tissue,
                             Sex_type = input$checkbox_sex,
                             Trimester_type = input$checkbox_trimester,
                             plot_heights = c(input$condensed_plot_height_1,
                                              input$condensed_plot_height_2,
                                              input$condensed_plot_height_3,
                                              input$condensed_plot_height_4))
      }
      #output a plot when input a region
      else if(input$tabs == 'Region Input')
        {
        annoPlot_with_tracks(cpg_name =  NULL, gene_name = NULL, 
                             Start = get_region_start(), 
                             End = get_region_end(), 
                             Chr = get_region_chr(), 
                             cpg_number = cpg_number_region$cpg_number, 
                             Tissue_type = input$checkbox_tissue,
                             Sex_type = input$checkbox_sex,
                             Trimester_type = input$checkbox_trimester,
                             plot_heights = c(input$condensed_plot_height_1,
                                              input$condensed_plot_height_2,
                                              input$condensed_plot_height_3,
                                              input$condensed_plot_height_4))
        }
      )
    })
    
    ############  annoplot output: choose a set of cpg (given a gene input)############
    #store cpg_start and cpg_end into reactive value
    cpg_start <- reactiveValues(cpg_start = 1)
    cpg_end <- reactiveValues(cpg_start = 500)
    #get the cpg set input
    observeEvent(input$submit_cpg_set, {cpg_start$cpg_start <- input$range_cpg[1]
                                        cpg_end$cpg_end <- input$range_cpg[2]})
    #reset cpg set when select another gene
    observeEvent(input$submit_gene_input, {cpg_start$cpg_start = 1
                                           cpg_end$cpg_end = 500})
    #update the plot when change the cpg number to show
    observeEvent(input$submit_cpg_set, {output$annoplot <- renderPlot(
      if (input_value$input_value  == FALSE) return()
      else if(input$tabs == 'Gene Input')
        {
        annoPlot_with_tracks(cpg_name = NULL, gene_name = get_gene_input(), 
                             Chr = NULL, Start = NULL, End = NULL, 
                             first_cpg = cpg_start$cpg_start, end_cpg = cpg_end$cpg_end, 
                             Tissue_type = input$checkbox_tissue,
                             Sex_type = input$checkbox_sex,
                             Trimester_type = input$checkbox_trimester,
                             plot_heights = c(input$condensed_plot_height_1,
                                              input$condensed_plot_height_2,
                                              input$condensed_plot_height_3,
                                              input$condensed_plot_height_4))
        
      }
      #output a plot when input cpgs
      else if(input$tabs == 'CpG Input')
        {
        annoPlot_with_tracks(cpg_name = get_cpg_input(), gene_name = NULL, 
                             Chr = NULL, Start = NULL, End = NULL, 
                             Tissue_type = input$checkbox_tissue,
                             Sex_type = input$checkbox_sex,
                             Trimester_type = input$checkbox_trimester,
                             plot_heights = c(input$condensed_plot_height_1,
                                              input$condensed_plot_height_2,
                                              input$condensed_plot_height_3,
                                              input$condensed_plot_height_4))
      }
      #output a plot when input a region
      else if(input$tabs == 'Region Input')
        {
        annoPlot_with_tracks(cpg_name =  NULL, gene_name = NULL, 
                             Start = get_region_start(), 
                             End = get_region_end(), 
                             Chr = get_region_chr(), 
                             first_cpg = cpg_start$cpg_start, 
                             end_cpg = cpg_end$cpg_end, 
                             Tissue_type = input$checkbox_tissue,
                             Sex_type = input$checkbox_sex,
                             Trimester_type = input$checkbox_trimester,
                             plot_heights = c(input$condensed_plot_height_1,
                                              input$condensed_plot_height_2,
                                              input$condensed_plot_height_3,
                                              input$condensed_plot_height_4))
        }
      )
    })
    
    
    
    ############  Expanded view output ############
    ############  (given a gene input) ############
    #gene cpg plots 
    output$expanded <- renderPlot(
      {
      #return empty when no inputs
      if (input_value$input_value  == FALSE) return()
        
      #output a plot when input a gene
      else if(input$tabs == 'Gene Input')
        {
        expanded_view(gene_name = input$gene_name, 
                
                start_extend = 2500, # how far to extend the plot left, in genomic coordinates 
                end_extend = 2500,   # how far to extend right
                show_transcript_gene = TRUE, # show transcripts associated gene symbol
                breaks_width = 15000, # determines the spacing of x axis breaks
                trimester = input$checkbox_trimester, # can be Third, First, or Both
                sex = input$checkbox_sex, # can be Male, Female, or Both
                tissue = input$checkbox_tissue,
                plot_sizes = c(input$expanded_plot_height_1,
                               input$expanded_plot_height_2))
      }
        }
      )
    
    ############  boxplot output ############
    ##retrive cpg name when click the datatable
    get_cpg_from_dt <- eventReactive(input$submit_cpg_gene_input, {input$cpg})
    
    #produce a boxplot based on the selected cpg from datatable
    observeEvent(input$dt_anno_cpg_cell_clicked, 
                 {
                   c_info = input$dt_anno_cpg_cell_clicked
                   
                   output$boxplot_selected_cpg <- renderPlot(
      {
        if (!is.null(c_info$value))
        {
          boxplot_selected(c_info$value, sex_type = input$checkbox_sex)
        }
        else if(!is.null(get_cpg_from_dt()))
        {
          boxplot_selected(get_cpg_from_dt(), sex_type = input$checkbox_sex)
        }
      }
    )
    })
    #produce a boxplot based on the selected cpg using search box
    observeEvent(input$submit_cpg_gene_input, {output$boxplot_selected_cpg <- renderPlot(
      {
        boxplot_selected(get_cpg_from_dt(), sex_type = input$checkbox_sex)
      }
    )})
    
    ############  reset boxplot output ############
    #reset the associated cpg inputs in search box
    observeEvent(input$reset_cpg_gene_input, {reset('cpg')})
    #reset the boxplot when click the reset button
    observeEvent(input$reset_cpg_gene_input, {output$boxplot_selected_cpg <- renderPlot(
      {
        return()
      }
    )})
    #reset the boxplot when switch between tabs
    observeEvent(input$tabs, {output$boxplot_selected_cpg <- renderPlot(
      {
        return()
      }
    )})
    #reset the boxplot when change into another gene
    observeEvent(input$submit_gene_input, {output$boxplot_selected_cpg <- renderPlot(
      {
        return()
      }
    )})
    #reset the boxplot when reset gene input
    observeEvent(input$reset_gene_input, {output$boxplot_selected_cpg <- renderPlot(
      {
        return()
      }
    )})
   
    
}
```

#Run App

```{r}
shinyApp(ui, server)
```
