---
title: "process_data"
author: "Victor Yuan"
date: "06/12/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

Some results / data is unnecessarily large. In the interest of saving computation time, and 
storage space, it will sometimes make sense to filter out unnecessary data, or process into more 
summarized / compressed forms. This script is to document these types of changes.

# Annotation

## Base

The annotation that we base our app on is a custom one that I (Victor) created using a couple different databases on packages. In brief, I used the Illumina-provided EPIC annotation for positional information on the hg19 genome build. Then, using the R package annotatr, I added information on cpg islands (UCSC), genes, transcripts (UCSC), and fantom5 enhancers. I 4 different imprinting studies to determine imprinted regions, which I describe in my cells publication. For PMDs, I used DI Schroeder et al. 2013 paper.

```{r}
library(tidyverse)
library(here)

# annotation
anno <- readRDS(file.path('Z:', 'Victor', 'Repositories', 'EPIC_annotation',
                          'hg19_epic_annotation.rds'))

# remove unnecessary columns
anno <- anno %>%
  select(-contains('width'), -contains('repeat'), -group, -genes_gene_id, -imprint_sources, 
         -imprint_methylated_allele, -imprint_region) %>%
  select(cpg:enhancers_id, genes_symbol, genes_tx_id, genes_id, everything()) # rearrange

```

## DMCS

Here, I process the differential methylation results, and remove unnecessary columns.

these are all DMCs that are significant at bonf < 0.01, and absolute difference beta value > 0.1. 

```{r}
dmcs <- readRDS(file.path("Z:", "Victor", "Projects", "NIH - cells", 
                          "data", "main", "interim", '2_4_dmcs.rds'))

dmcs <- dmcs %>%
  filter(bonferroni < 0.01, abs(delta_b) > 0.1) %>%
  select(gene, Group1, bonferroni, delta_b) %>%
  dplyr::rename(cpg = gene) %>%
  separate(Group1, c('Trimester', 'Tissue'), sep = '\\.') %>%
  mutate(Tissue = case_when(
    Tissue == 'Endo_cs' ~ 'Endothelial cs',
    Tissue == 'Hofb_cs' ~ 'Hofbauer cs',
    Tissue == 'Strom_cs' ~ 'Stromal cs',
    Tissue == 'Troph_cs' ~ 'Trophoblasts cs')) 

# widen
dmcs <- dmcs %>%
  mutate(sig = (bonferroni < 0.01) & (abs(delta_b) > 0.1)) %>%
  pivot_wider(id_cols = cpg,
              names_from = c(Trimester, Tissue),
              names_sep = '_',
              values_fill = list(sig = FALSE),
              values_from = sig)

# join to annotation
anno <- anno %>% 
  left_join(dmcs, by = 'cpg') %>% 
  mutate_if(is.logical, function(x)if_else(is.na(x), FALSE, x)) 

# separate out placenta and other imprints
anno <- anno %>%
  mutate(imprinted_dmr_placenta = imprint_tissue_specificity == 'placental-specific',
         imprinted_dmr_general = imprint_tissue_specificity == 'other') %>%
  select(-imprint_tissue_specificity) %>% 
  mutate_if(is.logical, function(x)if_else(is.na(x), FALSE, x)) 

#reorder
anno <- anno %>%
  select(cpg:genes_id, imprinted_dmr_general, imprinted_dmr_placenta, pmd_id, everything())
```

# pData and betas

```{r}
# sample data
pDat <- readRDS(here::here('data', '3_1_pDat_filt.rds'))
# methylation data
betas <- readRDS(here::here('data', '3_1_betas_filt.rds'))

# make sure all in same order
all(pDat$Sample_Name == colnames(betas))

set.seed(0)
pDat <- pDat %>% 
  mutate(deidentified_id = paste0('rob', sample(1:n(),n()))) %>%
  select(deidentified_id, Sex, Trimester, Tissue)

colnames(betas) <- pDat$deidentified_id
```

# summarize betas

to reduce memory size I reduce the size of the betas by summarizing into mean and sd

divide by sex trimester and cell type (30)
divide by trimester and cell type (10)


```{r}
betas_t <- t(betas) %>% 
      as_tibble %>% 
      mutate(deidentified_id = colnames(betas)) %>%
    
      # reshape into longer format
      pivot_longer(cols = -deidentified_id,
                   names_to = 'cpg', 
                   values_to = 'beta')  %>%
      
      # add tissue and trimester info
      left_join(pDat %>% select(deidentified_id, Trimester, Tissue, Sex), by = 'deidentified_id')

saveRDS(betas_t, here::here( 'data', 'betas_t.rds')) #5 gb

betas_sum <- betas_t %>%
  group_by(cpg, Trimester, Tissue) %>%
  summarize(mean =mean(beta),
            sd = sd(beta)) 

betas_sum_wide <- betas_sum %>%
  pivot_wider(id_cols = cpg,
              names_from = c('Trimester', 'Tissue'),
              values_from = c('mean', 'sd'))

betas_sum_wide_mat <- as.matrix(betas_sum_wide[,-1])
rownames(betas_sum_wide_mat) <- betas_sum_wide$cpg

object_size(betas_sum_wide_mat) #175 MB
object_size(betas) # down from 826 MB
```

# Save data

```{r eval = FALSE}
saveRDS(anno, here('data', 'annotation.rds'))
saveRDS(pDat, here('data', 'pDat.rds'))
saveRDS(betas, here('data', 'betas.rds'))
saveRDS(betas_sum_wide_mat, here('data', 'betas_summarized.rds'))
```

## IGV style plots

These plots for the expanded view requires transcript models. Here I use the R package annotatr to download these.

```{r}
library(annotatr)
#ect annotations for intersection with regions
# Note inclusion of custom annotation, and use of shortcuts
annots <- c('hg19_cpgs', 'hg19_basicgenes', 'hg19_genes_intergenic',
           'hg19_genes_intronexonboundaries')

# Build the annotations (a single GRanges object)
annotations <- build_annotations(genome = 'hg19', annotations = annots)
annotations <- annotations %>%
  as_tibble() %>%
  mutate(type = gsub('genes_', '',
                     gsub('hg19_', '', type)),
         type = fct_relevel(type,
                            c('1to5kb', 'promoters', '5UTRs', 
                              'exons', 'introns', 'intronexonboundaries', 
                              '3UTRs',
                              'cpg_islands', 'cpg_shores', 'cpg_shelves', 'cpg_inter'))) 
object.size(annotations)
```

# Save data

```{r eval = FALSE}
saveRDS(annotations, here('data', 'annotation_annotatr.rds'))
```
